### PR TITLE
FRONTEND-5167 :: Feature Internal :: Refactor Angular sample to use view-states

### PR DESCRIPTION
**Pull request by urijp**
**Merge / Pull request information:**
  - ViewState interface created in Harmony-TS core
  - Presenters deleted. View-states have been implemented in the two components we have.
  - Renamed components folder using the feature name.
  - Created business-logic folder that contains domain and data folders

** Asana task link:
https://app.asana.com/0/1109863238977521/1203218267235167/f

### CODE REVIEW

**Review by Alex-DA — APPROVED** [review]

**Review by joselufo — APPROVED** [review]

File: examples/angular/src/business-logic/domain/app.provider.ts [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import { CacheRepository, DefaultObjectValidator, GetInteractor, GetRepository, InMemoryDataSource, RepositoryMapper, SingleGetDataSourceRepository, VoidDataSource } from '@mobilejazz/harmony-core';
 
-import { HackerNewsStoryNetworkDataSource } from 'src/data/data-sources/hacker-news-story.network.data-source';
-import { HackerNewsStoryEntity } from 'src/data/entities/hacker-news-item.entity';
+import { HackerNewsStoryNetworkDataSource } from 'src/business-logic/data/data-sources/hacker-news-story.network.data-source';
```
Comment (lucianosantana) [inline examples/angular/src/business-logic/domain/app.provider.ts:3]: You can shorten the import path by using `../` instead of `src/business-logic/`

**Comment by lucianosantana** [comment]
[app.provider.module.ts](https://github.com/mobilejazz/harmony-typescript/pull/115/files#diff-957c86f1640a78aeb4902627764b967ac144c765921b0dbf1c845747faa00348) should live in the application folder.

**Comment by lucianosantana** [comment]
We said that inside business-logic folder we would have a feature sub-folder, and inside it data, domain and providers. Similar to the [mobile structure](https://harmony.mobilejazz.com/docs/best-practices/folders-structure#native-projects)

**Review by lucianosantana — CHANGES_REQUESTED** [review]

**Comment by doup** [comment]
> We said that inside business-logic folder we would have a feature sub-folder, and inside it data, domain and providers. Similar to the [mobile structure](https://harmony.mobilejazz.com/docs/best-practices/folders-structure#native-projects)

Btw, if the idea is to do something similar to mobile… can we rename `business-logic` to `features`?

I agree that inside that we should be putting feature folders, although, maybe in something simple like this it doesn't matter so much?

File: examples/angular/src/app/story/latest-ask-stories/latest-ask-stories.component.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+import { NavigationService } from "../../services/navigation.service";
+
+@Component({
+  selector: 'app-latest-ask-stories',
+  templateUrl: './latest-ask-stories.component.html',
+  styleUrls: ['./latest-ask-stories.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class LatestAskStoriesComponent implements OnInit {
+  public viewState!: LatestAskStoriesViewState;
```
Comment (doup) [inline examples/angular/src/app/story/latest-ask-stories/latest-ask-stories.component.ts:20]: **Nit:** Does it make sense to shorten this to `state`? Would it be confusing? For me better if it's terse, but then it might lose meaning.

**Comment by lucianosantana** [comment]
> > We said that inside business-logic folder we would have a feature sub-folder, and inside it data, domain and providers. Similar to the [mobile structure](https://harmony.mobilejazz.com/docs/best-practices/folders-structure#native-projects)
> 
> Btw, if the idea is to do something similar to mobile… can we rename `business-logic` to `features`?
> 
> I agree that inside that we should be putting feature folders, although, maybe in something simple like this it doesn't matter so much?

I don't mind naming the folder "features", but it leaves room for interpretation and some UI code might slip in there. 

Because this is the demo/reference project, I'd use the full-blown folder structure. We can't define a standard by starting with an exception to it.

File: packages/core/src/application/view-state.ts [hunk 3]
```diff
@@ -0,0 +1,3 @@
+export interface ViewState {
+    $type: Lowercase<string>;
```
Comment (doup) [inline packages/core/src/application/view-state.ts:2]:
```suggestion
    readonly $type: Lowercase<string>;
```

**Review by doup — APPROVED** [review]

**Comment by doup** [comment]
I would really like to find an easy way to enforce that the `ViewState`s are inmutable. Right now the options we have:

- mark component `viewState` as `Readonly<T>`?
- Wrap the state union type in `Readonly`?
- Add @Freeze decorator that freezes the class?

I don't like any of these, probably this should be a custom lint. When we detect anything that implements `ViewState`, then all its public props should be `readonly`. 🤔

File: examples/angular/src/app/app-routing.module.ts [hunk 4]
```diff
@@ -1,7 +1,7 @@
 import { NgModule } from '@angular/core';
 import { RouterModule, Routes } from '@angular/router';
-import { ShowStoryComponent } from './components/show-story/show-story.component';
-import { LatestAskStoriesComponent } from './components/latest-ask-stories/latest-ask-stories.component';
+import { ShowStoryComponent } from './story/show-story/story.component';
+import { LatestAskStoriesComponent } from './story/latest-ask-stories/latest-ask-stories.component';
 
 const routes: Routes = [
```
[the remark this answers is not in the document]
Comment (orioljp) [reply examples/angular/src/app/app-routing.module.ts:6]: @doup This PR has a scope, all new ideas must be in new tasks.

**Comment by orioljp** [comment]
I'm with @lucianosantana . I don't mind calling it features if we agree doing it like mobile team, but we have a clear definition about what it contains and features is not clear enough.

cc @franmontiel

**Comment by doup** [comment]
Btw, isn't `business-logic` synomym of `domain`? What I mean, does `business-logic` mean `domain+data`, or just `domain`?

**Comment by joselufo** [comment]
To add context, we use `features` as a library concept or a micro-service. 

`features` should be atomic, and if they need something from another, `feature` must be provided by the DI. So a feature contains at least the following structure:
* `{FEATURE_NAME}Provider.kt` --> Represent the primary DI for the feature, and it's the main public contract. 
* domain
* data (if needed, as you could be using generic datasources and repositories)

File: examples/angular/src/app/story/latest-ask-stories/latest-ask-stories.component.ts (hunk 2 above)
Comment (urijp) [reply examples/angular/src/app/story/latest-ask-stories/latest-ask-stories.component.ts:20]: yes, if we are making this project as a sample, I rather being strict with the naming

**Comment by urijp** [comment]
Changes I made:
- I renamed the business-logic folder into features
- I created the feature hacker-news-story
- The cache decorator has been moved inside core
- The providers inside features are now framework agnostic, this way we provide the providers but in a plane TS.

File: examples/angular/src/features/hacker-news-story/hacker-news-story.provider.ts [hunk 5]
```diff
@@ -1,25 +1,24 @@
...
+import { HackerNewsStoryNetworkDataSource } from 'src/features/hacker-news-story/data/data-sources/hacker-news-story.network.data-source';
+import { HackerNewsStoryEntity } from 'src/features/hacker-news-story/data/entities/hacker-news-item.entity';
+import { HackerNewsStoryJSONToHackerNewsStoryEntityMapper } from './data/mappers/hacker-news-story.mapper';
+import { HackerNewsFetchService, HackerNewsService } from './data/service/hacker-news.service';
+import { GetHackerNewsLatestAskStoriesInteractor } from './domain/interactors/get-hacker-news-latest-ask-stories.interactor';
+import { GetHackerNewsStoryInteractor } from './domain/interactors/get-hacker-news-story.interactor';
+import { HackerNewsStoryEntityToHackerNewsStoryMapper, HackerNewsStoryToHackerNewsStoryEntityMapper } from './domain/mappers/hacker-news-story.mapper';
+import { HackerNewsStory } from './domain/models/hacker-news-story.model';
+import { HackerNewsStoryIdsNetworkDataSource } from "./data/data-sources/hacker-news-story-ids.network.data-source";
+import {CacheDecoratorFactory} from "@bugfender/sdk/lib/utils";
```
Comment (doup) [inline examples/angular/src/features/hacker-news-story/hacker-news-story.provider.ts:11]: I don't think you wanted to import this. The import you want is:

```
import {createCacheDecorator} from '@mobilejazz/harmony-core';
```

And use it like:

```
const Cached = createCacheDecorator();
```

File: examples/angular/src/features/hacker-news-story/hacker-news-story.provider.module.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    provide: GetHackerNewsLatestAskStoriesInteractor,
+    deps: [HackerNewsStoryProvider],
+    useFactory: (provider: HackerNewsStoryProvider) => provider.getHackerNewsLatestAskStories(),
+  },
+  {
+    provide: GetHackerNewsStoryInteractor,
+    deps: [HackerNewsStoryProvider],
+    useFactory: (provider: HackerNewsStoryProvider) => provider.getHackerNewsStory(),
+  },
+];
```
Comment (doup) [inline examples/angular/src/features/hacker-news-story/hacker-news-story.provider.module.ts:24]: Since we're already implicitly coupled to Angular, why we don't just create an actual Angular module? or at least add a correct type to `HACKER_NEWS_PROVIDERS: Route[]` so that we get type checking.

Before someone else mentions this shouldn't be here. For me it's OK if we leave this file inside `features/<feature>/<feature>.provider.module.ts`; it's going to be much more practical than moving it to the `app` folder. Also, as long as the Angular coupling is limited to this file I'm fine with it.

**Review by doup — APPROVED** [review]
Please check the cache decorator import, you're using the one from Bugfender (which served as inspiration for the Harmony one).

File: examples/angular/src/features/hacker-news-story/hacker-news-story.provider.module.ts (hunk 6 above)
Comment (doup) [reply examples/angular/src/features/hacker-news-story/hacker-news-story.provider.module.ts:24]: What I mean with "implicitly coupled to Angular" is that this array follows the `Route[]` structure that Angular expects, so, even if we're not importing anything Angular related it is indeed coupled. That's why I'm proposing to go full in and use Angular types/imports inside this file.

If you really want to be decoupled, I suppose we could define our types and a mapper to convert to `Route[]`.

My vote is for coupling.

**Review by franmontiel — APPROVED** [review]

File: examples/angular/src/features/hacker-news-story/hacker-news-story.provider.ts (hunk 5 above)
Comment (orioljp) [reply examples/angular/src/features/hacker-news-story/hacker-news-story.provider.ts:11]: Sure! I'll change that, it was an automatic import. We should remove the one inside bugfender if possible

File: examples/angular/src/features/hacker-news-story/hacker-news-story.provider.module.ts (hunk 6 above)
Comment (orioljp) [reply examples/angular/src/features/hacker-news-story/hacker-news-story.provider.module.ts:24]: The actual type of this array is `Provider[]` which is also coupled to angular, so indeed, not having the explicit import doesn't meen we're not coupled, so if we're all ok with that, I'll leave it as an Angular module.

**Comment by orioljp** [comment]
I updated the PR restoring the feature module and changing some improbable imports. Please @lucianosantana check it again.

File: examples/angular/src/features/hacker-news-story/hacker-news-story.provider.module.ts (hunk 6 above)
Comment (doup) [reply examples/angular/src/features/hacker-news-story/hacker-news-story.provider.module.ts]: lol, sorry, yes `Provider[]`… I've done too much Libris router refactoring lately. 😅

File: examples/angular/src/features/hacker-news-story/hacker-news-story.provider.ts (hunk 5 above)
Comment (doup) [reply examples/angular/src/features/hacker-news-story/hacker-news-story.provider.ts]: Bugfender it's not exporting it from the root… I'm not sure if that can be fixed.

File: examples/angular/src/features/hacker-news-story/hacker-news-story.provider.module.ts (hunk 6 above)
Comment (doup) [reply examples/angular/src/features/hacker-news-story/hacker-news-story.provider.module.ts]: @lucianosantana what do you thing about this? I think it's the last controversial change. See: https://github.com/mobilejazz/harmony-typescript/blob/eecf7f3b73aae8608db76ed4d3993f55533b90f7/examples/angular/src/features/hacker-news-story/hacker-news-story.provider.module.ts

File: examples/angular/src/features/hacker-news-story/hacker-news-story.provider.module.ts [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+import { NgModule } from "@angular/core";
+
+@NgModule({
+  providers: [
+    {
+      provide: HackerNewsStoryProvider,
+      deps: [],
+      useFactory: () => {
+        return new HackerNewsStoryDefaultProvider();
+      },
```
Comment (doup) [inline examples/angular/src/features/hacker-news-story/hacker-news-story.provider.module.ts:14]: **Nit:**

```suggestion
      useFactory: () => new HackerNewsStoryDefaultProvider(),
```

File: examples/angular/src/features/hacker-news-story/hacker-news-story.provider.module.ts (hunk 6 above)
Comment (lucianosantana) [reply examples/angular/src/features/hacker-news-story/hacker-news-story.provider.module.ts]: Creating an Angular module or having it typed as provider both makes sense to me. What doesn't make sense is that this file lives in the feature folder. 

For me, the harmony/business logic part finishes with the providing of interactors [in this file](https://github.com/mobilejazz/harmony-typescript/blob/eecf7f3b73aae8608db76ed4d3993f55533b90f7/examples/angular/src/features/hacker-news-story/hacker-news-story.provider.ts). The way the app wants to consume those interactors (all being injected in a single module or DI split in different modules) is a decision to be made from the app's perspective. 

Specifically I'd move [hacker-news-story.provider.module.ts](https://github.com/mobilejazz/harmony-typescript/blob/eecf7f3b73aae8608db76ed4d3993f55533b90f7/examples/angular/src/features/hacker-news-story/hacker-news-story.provider.module.ts) to `/examples/angular/src/app/story`

Comment (doup) [reply examples/angular/src/features/hacker-news-story/hacker-news-story.provider.module.ts]: I agree in part with what you say; this file is Angular code, hence it should go to the `app` folder. That's the correct thing to do.

But, I don't think it's the practical thing. There might or not a relation between `src/app` and `features` folders; for example, as you suggest we are already moving it into the `story` folder which doesn't match with `HackerNewsStory` feature. What happens when different UI modules use that provider? Where should we put it?

Leaving it next to the pure TS provider is convenient and easy to understand (although, _incorrect_).

Another option maybe is having a `providers` folder, or something like that, where we can put these provider modules… at least is less confusing where to put the file when we might have several UI modules that use the same provider.

I'm leaned towards the convenience option; although I'm OK also if we find a good place for these inside `app`.

Comment (orioljp) [reply examples/angular/src/features/hacker-news-story/hacker-news-story.provider.module.ts]: As @lucianosantana well pointed out, this provider is coupled to Angular so features folder looks not to be the right place to place it, but as we discussed, the UI and Features don't match in terms of structure so the best approach would be having a providers folder.

The points agains doing this is that in terms of development we need to create a provider for every feature, so it's completely linked and having this splitted in two folders is so unconvenient. Until we agree on another approach, I'm leaving as it is.
